File: src/test/scala/com/exasol/cloudetl/bucket/AbstractBucketTest.scala
package com.exasol.cloudetl.bucket

import com.exasol.ExaConnectionInformation
import com.exasol.ExaMetadata
import com.exasol.cloudetl.storage.StorageProperties

import org.mockito.Mockito.when
import org.scalatest.BeforeAndAfterEach
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.mockito.MockitoSugar

class AbstractBucketTest extends AnyFunSuite with BeforeAndAfterEach with MockitoSugar {

  private[cloudetl] val PATH: String = "BUCKET_PATH"
  private[cloudetl] val FORMAT: String = "DATA_FORMAT"
  private[cloudetl] var properties: Map[String, String] = _

  override def beforeEach(): Unit = {
    properties = Map.empty[String, String]
    ()
  }

  protected[this] final def getBucket(params: Map[String, String]): Bucket =
    Bucket(StorageProperties(params))

  protected[this] final def getBucket(
    params: Map[String, String],
    exaMetadata: ExaMetadata
  ): Bucket =
    Bucket(StorageProperties(params, exaMetadata))

  protected[this] final def mockConnectionInfo(
    username: String,
    password: String
  ): ExaMetadata = {
    val metadata = mock[ExaMetadata]
    val connectionInfo: ExaConnectionInformation = new ExaConnectionInformation() {
      override def getType(): ExaConnectionInformation.ConnectionType =
        ExaConnectionInformation.ConnectionType.PASSWORD
      override def getAddress(): String = ""
      override def getUser(): String = username
      override def getPassword(): String = password
    }
    when(metadata.getConnection("connection_info")).thenReturn(connectionInfo)
    metadata
  }

  final def assertNoConnectionName(fn: => Unit): Unit = {
    val thrown = intercept[BucketValidationException] {
      fn
    }
    val message = thrown.getMessage()
    assert(message.contains("No CONNECTION_NAME property is defined"))
    assert(message.contains("Please use connection object to provide access credentials"))
    ()
  }

  final def assertForbiddenProperty(fn: => Unit): Unit = {
    val thrown = intercept[BucketValidationException] {
      fn
    }
    val message = thrown.getMessage()
    assert(message.contains("Using credentials as parameters is forbidded"))
    assert(message.contains("Please use an Exasol named connection object"))
    ()
  }

}

File: src/test/java/com/exasol/cloudetl/bucket/AbstractBucketTest.java
package com.exasol.cloudetl.bucket;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.util.Map;

import org.junit.jupiter.api.BeforeEach;

import com.exasol.*;
import com.exasol.cloudetl.storage.StorageProperties;

abstract class AbstractBucketTest {
    protected static final String PATH = "BUCKET_PATH";
    protected static final String FORMAT = "DATA_FORMAT";
    protected Map<String, String> properties;

    @BeforeEach
    void abstractBucketBeforeEach() {
        this.properties = Map.of();
    }

    protected final Bucket getBucket(final Map<String, String> params) {
        return Bucket.create(new StorageProperties(params));
    }

    protected final Bucket getBucket(final Map<String, String> params, final ExaMetadata exaMetadata) {
        return Bucket.create(new StorageProperties(params, exaMetadata));
    }

    protected final ExaMetadata mockConnectionInfo(final String username, final String password) {
        final ExaMetadata metadata = mock(ExaMetadata.class);
        final ExaConnectionInformation connectionInformation = new ExaConnectionInformation() {
            @Override
            public ConnectionType getType() {
                return ConnectionType.PASSWORD;
            }

            @Override
            public String getAddress() {
                return "";
            }

            @Override
            public String getUser() {
                return username;
            }

            @Override
            public String getPassword() {
                return password;
            }
        };
        try {
            when(metadata.getConnection("connection_info")).thenReturn(connectionInformation);
        } catch (final ExaConnectionAccessException exception) {
            throw new IllegalStateException(exception);
        }
        return metadata;
    }

    protected final void assertNoConnectionName(final ThrowingRunnable runnable) {
        final BucketValidationException thrown = assertThrows(BucketValidationException.class, runnable);
        final String message = thrown.getMessage();
        assertTrue(message.contains("No CONNECTION_NAME property is defined"));
        assertTrue(message.contains("Please use connection object to provide access credentials"));
    }

    protected final void assertForbiddenProperty(final ThrowingRunnable runnable) {
        final BucketValidationException thrown = assertThrows(BucketValidationException.class, runnable);
        final String message = thrown.getMessage();
        assertTrue(message.contains("Using credentials as parameters is forbidded"));
        assertTrue(message.contains("Please use an Exasol named connection object"));
    }

    @FunctionalInterface
    protected interface ThrowingRunnable extends org.junit.jupiter.api.function.Executable {
        @Override
        void execute();
    }
}
